UIE-92 Decouple Nav Router pt2

msilva-broad opened the pull request:
- implemented EnvironmentNavActions mechanics which decouple Environments component from terra-specific Nav implementation.
- EnvironmentNavActions type is used with NavLinkProvider generic to enforce type narrowing on nav arguments based on nav key parameter 1 of getUrl and navTo interface functions.
- EnvironmentsPage wrapper now provides a nav mapping implementation to satisfy NavLinkProvider by mapping the sanitized nav calls to terra-ui specific Nav calls.
- unit test coverage added.
- type improvements to Environments and related code that is close to the nav link arguments (name and namespace) being well typed as strings.
- nav.ts utils file first-pass conversion to Typescript, with the start of enum-like type safety for terra nav key names.

### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-92

### What
- should have no functional changes

### Why
- to support cleaner code sharing of Environments component with AoU team

### Testing strategy
- [ ] regression test that basic nav actions into Environments page and from it to terra-ui workspace-summary nav action work.

msilva-broad committed:
UIE-92 Decouple Nav Router pt2
- implemented EnvironmentNavActions mechanics which decouple Environments component from terra-specific Nav implementation.
- EnvironmentNavActions type is used with NavLinkProvider generic to enforce type narrowing on nav arguments based on nav key parameter 1 of getUrl and navTo interface functions.
- EnvironmentsPage wrapper now provides a nav mapping implementation to satisfy NavLinkProvider by mapping the sanitized nav calls to terra-ui specific Nav calls.
- unit test coverage added.
- type improvements to Environments and related code that is close to the nav link arguments (name and namespace) being well typed as strings.
- nav.ts utils file first-pass conversion to Typescript, with the start of enum-like type safety for terra nav key names.

dmohs commented on src/analysis/Environments/Environments.ts:
```diff
@@ -468,18 +474,22 @@ export const Environments: React.FC<EnvironmentsProps> = ({ nav = undefined }) =
-            { href: nav.getLink('workspace-dashboard', { namespace, name }), style: { wordBreak: 'break-word' } },
-            [name]
-          ),
+          h(Link, { href: nav.getUrl('view-workspace', { namespace, name }), style: { wordBreak: 'break-word' } }, [
```
Nice. getUrl is a more accurate name than getLink.

"view-workspace" doesn't make sense, though. A URL points to a resource, which is a noun. "workspace-dashboard" or "workspace-details" are nouns, but "view-workspace" is a verb, so there is a type mismatch here.

msilva-broad replied:
ah, I didn't realize the "noun" thinking here.  How about "workspace-view"?

msilva-broad replied:
fixed

dmohs commented on packages/core-utils/src/nav/nav-utils.ts:
```diff
@@ -0,0 +1,27 @@
+ * // and similar for navTo
+ */
+export interface NavLinkProvider<NavTypes> {
+  navTo: KeyedNavFn<NavTypes, void>;
```
I don't quite see why `navTo` is part of this interface. It seems to me that a generic `goToUrl(string)` would be enough.

nawatts replied:
We'd want components to declare what routes they might attempt to navigate to so that consumers can ensure that their application can handle those navigation requests. I think a generic `goToUrl(string)` would require a consumer to read the source code to understand all URLs that the component might try to navigate to.

Also, URL instead of route name means the application's route names become bound to the component's. Theoretically, an application should be able to name the route `/envs` even if the component calls it 'environments'.

msilva-broad replied:
As we discussed, I can add some thinking on this to the interface here.   NavTo was chosen to fully remove the "concern" of navigation and any side-effect of that that different nav solutions may need to handle.

msilva-broad replied:
documentation tweaked

nawatts replied:
I get why we renamed `getLink` to `getUrl` (it's a more descriptive name).

But does the name of this method need to diverge so much from existing code? I would prefer something like `goToRoute`: the `go` to mirror the names in Terra UI nav and the [History](https://developer.mozilla.org/en-US/docs/Web/API/History) API, and `route` to clarify that this doesn't take a URL as one might assume for a navigation API.

msilva-broad replied:
my thinking is even route is an implementation detail... the generic concern is "navigation"...

nawatts replied:
Fair enough. Though the comment about `goTo` vs `navTo` for more consistency with existing code still applies.

msilva-broad replied:
in usage, `nav.goTo(key) `seems fine... but I kinda like that if consumer deconstructs `const { getUrl, navTo } = nav;` the names are more descriptive for later use in the component code.

`navTo(key)`

nawatts replied:
I'll claim that the improvement in readability is based on an assumption about the consuming code destructuring the nav object. If it doesn't destructure, then I'd claim `nav.goTo` is slightly more readable than the repetitive `nav.navTo`. Also, destructuring allows renaming the variable, so a consumer could also do something like `const { goTo: navTo } = nav` if they wanted.

My objection is more about the break with convention. Existing Terra UI code uses `goToPath` and, when in doubt, I'd tend to default to following precedent since I think the familiarity aids understanding. I can see the argument for removing "path" (your point that it's an implementation detail). But the reasoning for changing "goTo" to "navTo" seems more tenuous.

But it's a fine enough point that I don't want to hold up the PR over it.

dmohs commented on src/analysis/Environments/Environments.test.ts:
```diff
@@ -78,8 +70,9 @@ const listRuntimesV2: () => Promise<ListRuntimeItem[]> = jest.fn();
-const defaultNav = {
-  getLink: jest.fn().mockReturnValue(''),
+const defaultNav: NavLinkProvider<EnvironmentNavActions> = {
+  getUrl: jest.fn().mockReturnValue(''),
```
This doesn't seem like a valid default implementation. I might have this throw instead to make that clear.

nawatts replied:
I think throwing here would cause the test to fail. But the mock could probably be more realistic if it returned `/environments` or something like that instead of an empty string.

msilva-broad replied:
just to clarify, this is just a mock for the unit tests.  I can rename defaultNav to mockNav, and have it return '/' to look like a valid route.

msilva-broad replied:
fixed to return '/' and renamed to mockNav

dmohs commented on src/libs/nav.test.ts:
```diff
@@ -0,0 +1,19 @@
+    const result = isTerraNavKey(terraNavKey('workspace-dashboard'));
+
+    // Assert
+    expect(result).toBe(true);
```
How could this pass given the changes?

msilva-broad replied:
I think we cleared this up.  Environments nav actions are purposefully named generically, and these terra key names are not changing.

dmohs commented on src/libs/nav.ts:
```diff
@@ -7,6 +7,38 @@ import { useOnMount, useStore } from 'src/libs/react-utils';
+ *   Typescript currently allows unsafe object['not-a-enum-key-name'] indexing
+ *   lookups into this otherwise type-safe construct.
+ */
+export const terraNavKeyValue: TerraNavKeyLookup = {
+  billing: 'billing',
+  'workspace-dashboard': 'workspace-dashboard',
+  workspaces: 'workspaces',
+};
```
I would construct this differently to make the intent clearer and reduce the possibility of typos:
```suggestion
export const terraNavKeyValue: TerraNavKeyLookup = [
  'billing', 'workspace-dashboard', 'workspaces'
].reduce((acc, k) => ({...acc, [k]: k}), {});
```

msilva-broad replied:
love it!

msilva-broad replied:
actually, I misread the suggestion... missed the reduce at the end... Typescript will prevent typos already (TerraNavKeyLookup insures that only valid keys can be used, and that the value must match the key).   While this suggestion was attractive because it removed the key: key repetition in the main body of the object, the reduce mechanic causes new Typescript problems since it's a bit too dynamic to guarantee the end result of .reduce is the full set of keys.  I got TS complaints when I tried this locally, and the type protection seemed to actually suffer in general.

blakery replied:
I can find some ways of doing it generically, but they're all way more complex and opaque than just writing them out.

dmohs commented on src/libs/nav.ts:
```diff
@@ -128,3 +160,13 @@ export function PathHashInserter() {
+
+export interface TerraNavLinkProvider {
+  getLink: (name: string, params?: Record<string, any>) => string;
+  goToPath: (name: string, params?: Record<string, any>) => void;
```
I am confused as to why these names are different. I may need some in-person help to grok this.

msilva-broad replied:
as we discussed, this is a mapping interface to help the current nav implementation in terra-ui be mappable to the new more generic NavLinkProvider interface, but still honor the existing method names.  We may iterate later and add more TS goodness, but this is meant as a babystep.

dmohs reviewed (commented):
First pass with a few comments. I'm hoping we can go through them in-person so I can learn.

nawatts commented on src/libs/nav.ts:
```diff
@@ -7,6 +7,38 @@ import { useOnMount, useStore } from 'src/libs/react-utils';
+};
+
+/**
+ * type guard for nav keys to better enforce key names with snake-case
```
This comment seems like it goes with `isTerraNavKey`. Also, how does this relate to snake-case keys?

msilva-broad replied:
good  catch, this missed my cleanup pass.

nawatts commented on src/libs/nav.ts:
```diff
@@ -7,6 +7,38 @@ import { useOnMount, useStore } from 'src/libs/react-utils';
+ * type guard for nav keys to better enforce key names with snake-case
+ * @param key
+ */
+export const terraNavKey = (key: TerraNavKey): TerraNavKey => key;
```
What is this function for?

msilva-broad replied:
this extra type guard fn is needed because 'the-nav-key', devs can't use dot notation to access the enum-constants, so this pass-thru helps provide the same safety.

msilva-broad replied:
nick: type guard is official terminonogy for what isTerraNavKey fn is below (it returns a type predicate).... hence the confusion...

how about we create a generic `enforceType<T>(value: T> => T `and call it here in terraNavKey.

nawatts replied:
`enforceType` sounds good to me.

Here's the relevant TS docs: https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types

An example usage might help here, showing how this can be used to require a narrower type than what the function that its return value will be passed to requires.

msilva-broad replied:
fixed

msilva-broad commented on src/pages/EnvironmentsPage.ts:
```diff
@@ -1,14 +1,42 @@
+  const myNav: NavLinkProvider<EnvironmentNavActions> = {
+    getUrl: <K extends keyof EnvironmentNavActions>(navKey: K, navArgs: EnvironmentNavActions[K]): string => {
+      const navMap: NavMap<EnvironmentNavActions, string> = {
+        'view-workspace': (args) => getLink(workspaceNavKey, args),
```
maybe switch this to inline terraNavKey call?

msilva-broad replied:
fixed

nawatts commented on packages/core-utils/src/nav/nav-utils.ts:
Is navigation a larger enough concern to warrant it's own package?

I could imagine one or both of...
a) Moving most/all of libs/nav into a nav package
b) Our approach to navigation changing more rapidly than core-utils

msilva-broad replied:
right now the footprint is pretty small, so I am inclined to keep it in core-utils for the time being....

nawatts commented:
Alternatively, we could completely remove the nav concern from Environments by having it take a function to render an environment's workspace's name. Consumers could implement that function to render a link to the workspace dashboard in the consuming app.

msilva-broad committed:
UIE-92 Decouple Nav Router pt2
- initial PR feedback fixes

msilva-broad commented:
> Alternatively, we could completely remove the nav concern from Environments by having it take a function (or component) to render an environment's workspace's name. Consumers could implement that function to render a link to the workspace dashboard in the consuming app.

Thinking that thru a bit.... hmm.... I think this would burden the shared component (Environments here) with url specifics and knowledge that don't belong to it, and make it harder for an alternate app consumer to change.... /path/to/workspace-view/[WorkspaceName] doesn't feel like something we want Environments component to have to craft.  (Also, the link needs workspace name and namespace string)

nawatts commented:
What I was thinking is that Environments takes a prop `renderWorkspaceName: ({ namespace, name }) => ReactNode`. It could have a default implementation of that props that renders the workspace name as text `({ name }) => name` or something like that.

EnvironmentsPage could then render Environments as:
```ts
h(Environments, {
  ...
  renderWorkspaceName: ({ namespace, name }) => h(Link, { href: Nav.getLink(...) }, [name]),
```

That way, the word 'nav` doesn't appear at all in Environments.

However, I think that approach could work well for Environments because it only links to one place. In a component that links to multiple routes, the props interface could get cluttered compared to a single nav object.

nawatts commented on src/libs/nav.ts:
```diff
@@ -7,6 +8,38 @@ import { useOnMount, useStore } from 'src/libs/react-utils';
+};
+
+/**
+ * type helper for nav keys to better enforce key names with snake-case
```
Quibble about this... snake-case keys aren't necessarily problematic. It's related to the type of the key used to access the value together with our TS configuration to allow implicit anys.

```ts
type Foo = {
    'snake-case-key': number,
}

const foo: Foo = {
    'snake-case-key': 1,
}

// literal key, n1 is type number
const n1 = foo['snake-case-key']

// key is type "snake-case-key", n2 is type number
const key = 'snake-case-key'
const n2 = foo[key]

// key is type string, n3 is type any
const key2: string = 'snake-case-key'
const n3 = foo[key2]
```

msilva-broad replied:
good to know... so later down the line when we are at 100% ts adoption and flip off the "allow any" switch, we can tighten things up.  That will take some time though, so these extra precautions are valuable in the meantime :)

nawatts commented on packages/core-utils/src/type-utils/type-helpers.ts:
```diff
@@ -3,3 +3,5 @@
  */
 export type NominalType<BaseType, Name extends string> = BaseType & { __typeToken: Name };
+
+export const enforceType = <T>(value: T): T => value;
```
It would be nice to add a doc string for this function since (in my opinion, at least) it's not clear where this would be useful.

msilva-broad replied:
will do

msilva-broad replied:
in response to related comments, I removed enforceType util and cleaned up terraNavKey and related documentation in nav.ts

nawatts commented on src/libs/ajax/leonardo/models/core-models.ts:
```diff
@@ -15,7 +15,7 @@ export interface LeoError {
 }
 
-export type LeoResourceLabels = Record<string, any>;
+export type LeoResourceLabels = { [key: string]: string };
```
Is there a difference between `Record<string, string>` and `{ [key: string]: string }`? I thought that in the past we had encouraged the Record type?

blakery replied:
They are technically equivalent, but the the record type is generally used when you want to define a specific set of keys, rather than a generic type of key. 

For example:
`type KeyedObjectType = Record<'key1' | 'key2' | 'key3', string>`
is the same as:
`type KeyedObjectType = {key1: string, key2: string, key3: string}

Generally speaking the index signature (`{ [key: string]: string }`) is more idiomatic when you have the generic, unrestricted keys, like here.

This is just the general ts idiom though, so it's possible we decided on something else.

nawatts commented on src/libs/nav.ts:
```diff
@@ -7,6 +8,38 @@ import { useOnMount, useStore } from 'src/libs/react-utils';
+ * type helper for nav keys to better enforce key names with snake-case
+ * @param key
+ */
+export const terraNavKey = (key: TerraNavKey) => enforceType<TerraNavKey>(key);
```
```suggestion
export const terraNavKey = enforceType<TerraNavKey>;
```

This could be simplified. I find the simpler version clearer. As is, there's no reason to use `enforceType` because the type is already enforced by the type declaration on the `key` parameter.

msilva-broad replied:
I hit an ergonomic barrier where just using enforceType didn't give the IDE enough direct TS goodness to fully restrict the auto-suggest values of key arg to the valid key values... so I had to re-introduce the arg typing here... I kinda like that enforceType is self-documenting even if the end result here is not as crisp/concise as we hoped.

nawatts replied:
In that case, I still think the `enforceType` is useless here: it's an extra function call for no gain. I'd vote for something like:
```suggestion
export const enforceTerraNavKeyType = (key: TerraNavKey) => key;
```

nawatts replied:
Looks like it depends on the editor too... VS Code autocompletes arguments for something like `const enforceMyType = enforceType<MyType>;`.

nawatts commented on src/pages/EnvironmentsPage.ts:
```diff
@@ -1,14 +1,40 @@
+  [Property in keyof NavTypes]: (args: NavTypes[Property]) => FnReturn;
+};
+
+export const makeNavProvider = (terraNav: TerraNavLinkProvider): NavLinkProvider<EnvironmentNavActions> => {
```
Not necessarily needed now, but it would be nice to have a helper to construct these objects. It looks like, in most cases, these nav providers could be constructed from a map of component route names to Terra UI route names (`workspace-view` => `workspace-dashboard` in this case).

msilva-broad replied:
agreed,  I saw the same eventuality.  But I left it as a potential TODO once n is 2 or 3.

nawatts commented on packages/core-utils/src/nav/nav-utils.ts:
```diff
@@ -0,0 +1,36 @@
+ *
+ * // and similar for navTo
+ */
+export interface NavLinkProvider<NavTypes> {
```
Maybe "NavProvider" instead of "NavLinkProvider". Since there's no longer a `getLink` method, the "link" in the name here seems out of place.

msilva-broad replied:
I avoided NavProvider since I am thinking that will be a broader interface with additional app-level utils like we see in Terra's nav.ts module.  I could go for NavActionProvider maybe?

nawatts replied:
Eh, but then `getUrl` isn't really an action. I'm not entirely convinced there's a conflict in names: I could see an ApplicationNavProvider that extends the basic NavProvider interface presented to components.

But also, Environments is a pretty high-level component. What would consume that application level interface if not something like Environments?

msilva-broad replied:
I am guided by the I in SOLID (Interface Segregation) which says to prefer narrow interfaces over broad ones.  This is a narrow interface just for nav-link functionality... so I feel a more specific name makes sense.  NavLinkProvider seems a good fit following that reasoning.  NavProvider is a broader, less specific name.

nawatts replied:
"Link" makes me think `<a href=...`. This is broader than that, since it allows navigation from any code, not only links, via the `navTo` function.

nawatts reviewed (commented):
A few comments, but the only thing stopping me from clicking "Approve" is the one about the name of the `navTo` method.

msilva-broad commented:
> What I was thinking is that Environments takes a prop `renderWorkspaceName: ({ namespace, name }) => ReactNode`. It could have a default implementation of that props that renders the workspace name as text `({ name }) => name` or something like that.
> 
> EnvironmentsPage could then render Environments as:
> 
> ```ts
> h(Environments, {
>   ...
>   renderWorkspaceName: ({ namespace, name }) => h(Link, { href: Nav.getLink(...) }, [name]),
> ```
> 
> That way, the word 'nav` doesn't appear at all in Environments.
> 
> However, I think that approach could work well for Environments because it only links to one place. In a component that links to multiple routes, the props interface could get cluttered compared to a single nav object.

Ok, yes... this type of re-composition is quite powerful, but I call YAGNI for now.  I am very motivated to keep public interfaces as narrow as possible since they are high in pain if and when we need to make breaking changes that affect multiple teams.

nawatts commented:
I agree with the conclusion, but my reasoning is basically the opposite... A render prop would be a narrower interface than NavLinkProvider, but we are eventually going to want something such as NavLinkProvider that generalizes more easily to support components that have more links and/or link to multiple other routes.

msilva-broad committed:
UIE-92 Decouple Nav Router pt2
- removed enforceType util and improved documentation for terraNavKey and isTerraNavKey

nawatts commented on src/libs/nav.ts:
```diff
@@ -83,15 +121,15 @@ export const getCurrentRoute = () => {
 
 export const useRoute = () => {
-  const location = useContext(locationContext);
+  const location: any = useContext(locationContext);
```
```suggestion
  const location = useContext(locationContext);
```

Do we need the explicit `any`? It looks like the type of this would still currently be `any` without it (because `locationContext` is typed `Context<any>`). But when we improve the type for `locationContext`, we'd have to remember/notice this and remove it to get the benefit of that type here.

msilva-broad replied:
once we give nav.ts more Typescript love in general, things should click in better.  But i was running into type flow problems where `Context<any>` was not enough with the current state of related types and usage.

msilva-broad committed:
UIE-92 Decouple Nav Router pt2
- rebased with dev
